### DISCUSSION TITLE
Question about using adaptive mesh on Level-Set

### DISCUSSION THREAD

**Discussion by doquang** [idaholab/moose#23279] [Q&A Meshing] [answered]
Dear Moose Team,

I'm trying to use Level Sets module and I have a question about model convergence with "some kind" of mesh refinements through the level set interface.

Please see attached input file for detailed information. Note that, if you disable Adaptivity, the model converge to a correct solution! Or if you disable option #2, and enable option #1, the model will show a correct solution!

Thank you very much for your help,
/ Minh

```
size = 1
# Option 1: Adaptive mesh over all geometry size. Solution is ok
#asize = ${size}

# Option 2: Adaptive mesh over 1/3 geometry size. Solution is not ok
asize = ${fparse size/3}
[Mesh]
  type = GeneratedMesh
  dim = 3
  xmin = 0
  xmax = ${size}
  ymin = 0
  ymax = ${size}
  zmin = 0
  zmax = ${size}
  nx = 10
  ny = 10
  nz = 10
[]

[Variables]
  [./phi]
  [../]
[]

[AuxVariables]
  [./velocity]
    family = LAGRANGE_VEC
  [../]
[]

[ICs]
  [./phi_ic]
    type = FunctionIC
    function = phi_exact
    variable = phi
  [../]
  [./vel_ic]
    type = VectorFunctionIC
    variable = velocity
    function = velocity_func
  []
[]

[Functions]
  [./phi_exact]
    type = LevelSetOlssonPlane
    point = '${fparse size/2} ${fparse size/2} ${fparse size/2}'
    normal = '0 0 1'
    epsilon = ${fparse size/15}
  [../]
  [./velocity_func]
    type = ParsedVectorFunction
    expression_x = '1e-15'
    expression_y = '1e-15'
    expression_z = '1e-15'
  [../]
[]

[BCs]
  [top]
    type = DirichletBC
    variable = phi
    boundary = front
    value = 1
  []
  [bottom]
    type = DirichletBC
    variable = phi
    boundary = back
    value = 0
  []
[]

[Kernels]
  [./advection]
    type = LevelSetAdvection
    velocity = velocity
    variable = phi
  [../]
  [./advection_supg]
    type = LevelSetAdvectionSUPG
    velocity = velocity
    variable = phi
  [../]
[]

[Adaptivity]
  initial_marker = initial_marker_phi
  initial_steps = 2
  max_h_level = 2
  cycles_per_step = 2
  recompute_markers_during_cycles = true

  marker = initial_marker_phi
  steps = 1

  [Markers]
    [initial_marker_phi]
      type = BoxMarker
      bottom_left = '0 0 ${fparse size/2.5}'
      top_right = '${size} ${asize} ${fparse size/2}'
      inside = refine
      outside = do_nothing
    []
  []
[]

[Executioner]
  type = Steady
  nl_rel_tol = 1e-3
  nl_abs_tol = 1e-7
  solve_type = NEWTON
  petsc_options_iname = '-pc_type -pc_factor_mat_solver_package'
  petsc_options_value = 'lu       superlu_dist'
[]

[Outputs]
  csv = true
  exodus = true
[]
```

**Comment by doquang** [comment]
Correct solution
<img width="680" alt="image" src="https://user-images.githubusercontent.com/3186923/215585949-7589934b-35f7-4755-a8e5-92ffafad8d85.png">

Wrong solution
<img width="681" alt="image" src="https://user-images.githubusercontent.com/3186923/215585710-b2f9974a-62f3-4dd3-90dc-49d744ce3a90.png">

**Reply by GiudGiud** [reply]
Are these results converged? Can you paste the execution log for the second case

**Reply by doquang** [reply]
```
Framework Information:
MOOSE Version:           git commit 9721739139 on 2023-01-25
LibMesh Version:         a476dcb36d78923b9d2ed13b937b388f00da77f4
PETSc Version:           3.16.6
SLEPc Version:           3.16.2
Current Time:            Mon Jan 30 21:29:49 2023
Executable Timestamp:    Tue Jan 24 13:56:40 2023

Parallelism:
  Num Processors:          1
  Num Threads:             1

Mesh: 
  Parallel Type:           replicated
  Mesh Dimension:          3
  Spatial Dimension:       3
  Nodes:                   1331
  Elems:                   1000
  Num Subdomains:          1

Nonlinear System:
  Num DOFs:                1331
  Num Local DOFs:          1331
  Variables:               "phi" 
  Finite Element Types:    "LAGRANGE" 
  Approximation Orders:    "FIRST" 

Auxiliary System:
  Num DOFs:                4993
  Num Local DOFs:          4993
  Variables:               "velocity" "initial_marker_phi" 
  Finite Element Types:    "LAGRANGE_VEC" "MONOMIAL" 
  Approximation Orders:    "FIRST" "CONSTANT" 

Execution Information:
  Executioner:             Steady
  Solver Mode:             NEWTON
  MOOSE Preconditioner:    SMP (auto)

 0 Nonlinear |R| = 1.662917e-01
      0 Linear |R| = 1.662917e-01
      1 Linear |R| = 4.558296e-13
 1 Nonlinear |R| = 4.197392e-13
 Solve Converged!
Mesh unchanged, skipping remaining steps...
 0 Nonlinear |R| = 2.624513e-19
 Solve Converged!
```

**Reply by GiudGiud** [reply]
Seems to be well converging, much below the tolerances set

**Reply by GiudGiud** [reply]
Any update on this?

**Reply by doquang** [reply]
No, 
I still can not simulate a steady-state of a level-set problem with a different mesh size (adaptive mesh level) over the level-set interface.

**Reply by GiudGiud** [reply]
ok thanks for the update.

At this point I think this is because the problem is ill-posed. 
You have Dirichlet BCs on both sides yet are advecting from one face to the other. What do you think the steady state would look like?
How does phi go from 1 to 0 with just an advection (with a fixed velocity) process

**Reply by doquang** [reply]
Thanks @GiudGiud !

My problem occurs in the first stage of the simulation, when the velocity is very small. And it worked for me with the above BC diriclet, for a simple mesh (option #1).

By removing the BCs, I got a convergence for option #2, but it converged to a linear solution. Do you think it would be better using re-initialization?

<img width="675" alt="image" src="https://user-images.githubusercontent.com/3186923/216777641-c962c5dc-8632-4b63-93b3-2789823593a8.png">

**Reply by GiudGiud** [reply]
If the system is ill-posed the solution will depend on the mesh but is not really to be trusted.

Linear seems odd as well.
Dont you expect the value to just propagate through the entire mesh from the dirichlet BC?

**Reply by doquang** [reply]
I just expected, with a very small velocity, it should not convect the level-set, but it seems doesn't work for me, for the Steady state solver. Now I am testing my problem on a Transient system instead. 

Thank you very much!

**Reply by GiudGiud** [reply]
Transient seems like a good idea.

In steady even a tiny velocity will be able to advect across the domain, the timestep is essentially 'infinite'

**Reply by jiangwen84** [reply]
I agree that you should change to a transient problem. If it does not work, let us know. I have a few other Ideas why it does not work.

**Comment by GiudGiud** [comment, marked as answer]
nl_rel_tol = 1e-3
  nl_abs_tol = 1e-7 

this seems much too loose.
You need to tighten both

**Reply by doquang** [reply]
Yes, I will. 
This is just for testing
